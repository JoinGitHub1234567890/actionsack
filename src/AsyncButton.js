import { memo, useState } from 'react'

const useAsyncCallback = function(cb, initiallyLoading = false) {
  const [loading, setLoading] = useState(initiallyLoading)
  const [error, setError] = useState(null)

  return {
    loading,
    error,
    onAction: async function(...props) {
      setLoading(true)
      try {
        return await cb(...props)
      } catch (e) {
        setError(e)
      } finally {
        setLoading(false)
      }
    }
  }
}

export const AsyncButton = memo(function AsyncButton(props) {
  const { loading, error, onAction } = useAsyncCallback(props.onClick, props.initiallyLoading)
  return props.children({
    loading,
    error,
    onClick: onAction
  })
})
