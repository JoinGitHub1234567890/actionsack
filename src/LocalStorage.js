import React from 'react'

const toString = value => JSON.stringify(value)

export function useLocalStorage(key) {
  const [state, setState] = React.useState(null)

  React.useEffect(() => {
    if (localStorage.hasOwnProperty(key)) {
      // get the key's value from localStorage
      let value = localStorage.getItem(key)

      // parse the localStorage string and setState
      try {
        value = JSON.parse(value)
      } finally {
        setState(value)
      }
    }
  }, [key])

  React.useEffect(() => {
    function handleStorageChange(e) {
      const { isTrusted, newValue: value } = e

      if (isTrusted) {
        setState(value)
      }
    }

    window.addEventListener('storage', handleStorageChange)

    return () => window.removeEventListener('storage', handleStorageChange)
  }, [])

  React.useEffect(() => {
    if (state == undefined) {
      localStorage.removeItem(key)
    } else {
      localStorage.setItem(key, toString(state))
    }
  }, [key, state])

  return [state, setState]
}
