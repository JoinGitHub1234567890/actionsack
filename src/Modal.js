import React from 'react';
import enhanceWithClickOutside from 'react-click-outside';

const ESCAPE_KEY = 27;

const ClickAway = enhanceWithClickOutside(
  class extends React.Component {
    constructor(props) {
      super(props);
      this.escFunction = this.escFunction.bind(this);
    }
    handleClickOutside(e) {
      this.props.onClickAway(e);
    }
    escFunction(event) {
      if (event.keyCode === ESCAPE_KEY) {
        this.props.onClickAway();
      }
    }
    componentDidMount() {
      document.addEventListener('keydown', this.escFunction, false);
    }
    componentWillUnmount() {
      document.removeEventListener('keydown', this.escFunction, false);
    }

    render() {
      return this.props.children;
    }
  }
);

const Modal = props => (
    <ClickAway onClickAway={props.onClickAway}>
      {props.open ? props.children : null}
    </ClickAway>
  );

export default Modal;
