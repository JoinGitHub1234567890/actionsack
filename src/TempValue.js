import React from 'react'

function noop() {}
class TempValue extends React.Component {
  constructor(props) {
    super(props)

    this.base = {
      error: null,
      loading: false
      value: this.props.value
    }
    this.state = this.base

    this.handleChange = this.handleChange.bind(this)
    this.handleInputChange = this.handleInputChange.bind(this)
    this.handleCancel = this.handleCancel.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(value) {
    this.setState({ value })
  }

  handleInputChange(e) {
    this.handleChange(e.target.value)
  }

  handleCancel() {
    this.setState(this.base, this.props.onCancel)
  }

  async handleSubmit() {
    this.setState({ loading: true, error: null })
    try {
      await this.props.onSubmit(this.state.value)
      this.base = this.state
    } catch (error) {
      this.setState({ error })
    } finally {
      this.setState({ loading: false })
    }
  }

  render() {
    return this.props.children({
      value: this.state.value,
      loading: this.state.loading,
      error: this.state.error,
      onChange: this.handleChange,
      onInputChange: this.handleInputChange,
      onCancel: this.handleCancel,
      onSubmit: this.handleSubmit
    })
  }
}

TempValue.defaultProps = {
  onCancel: noop,
  onSubmit: noop
}

export default TempValue;
