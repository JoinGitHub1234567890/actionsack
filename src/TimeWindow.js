import React from 'react'

export function useTimeWindow({ showAfter, showUntil }) {
  const [show, setShow] = React.useState(showAfter == null)

  React.useEffect(() => {
    if (showAfter != null) {
      const timeoutId = setTimeout(setShow, showAfter, true)
      return () => clearTimeout(timeoutId)
    }
  }, [showAfter])

  React.useEffect(() => {
    if (showUntil != null) {
      const timeoutId = setTimeout(setShow, showUntil, false)
      return () => clearTimeout(timeoutId)
    }
  }, [showUntil])

  return show
}

export function TimeWindow({ children, showAfter, showUntil }) {
  const show = useTimeWindow({ showAfter, showUntil })

  if (!show) {
    return null
  }

  return children
}
