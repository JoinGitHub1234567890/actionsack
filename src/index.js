export * from './AsyncButton'
export * from './CopyButton'
export * from './Modal'
export * from './TempValue'
export * from './LocalStorage'
