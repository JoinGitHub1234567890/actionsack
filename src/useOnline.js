import React from 'react'

function useWindowListener(key, fn) {
  const callbackRef = React.useRef(fn)

  React.useEffect(() => {
    const cb = callbackRef.current
    window.addEventListener(key, cb)
    return () => window.removeEventListener(key, cb)
  }, [key])
}

export function useOnline() {
  const [online, setOnline] = React.useState(true)

  React.useEffect(() => {
    setOnline(
      typeof navigator !== 'undefined' && typeof navigator.onLine === 'boolean'
        ? navigator.onLine
        : true
    )
  }, [])

  useWindowListener('offline', () => setOnline(false))
  useWindowListener('online', () => setOnline(true))

  return online
}
