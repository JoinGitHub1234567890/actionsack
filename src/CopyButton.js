import React, { memo } from 'react'
import CopyToClipboard from 'react-copy-to-clipboard'

import { useAsyncCallback } from './AsyncButton'

const wait = t => () => new Promise(res => setTimeout(res, t))

export const CopyButton = memo(function CopyButton(props) {
  const [onCopy, { loading: copied }] = useAsyncCallback(
    wait(props.interval == null ? 1000 : props.interval)
  )

  return (
    <CopyToClipboard text={props.text} onCopy={onCopy}>
      {props.children({ copied })}
    </CopyToClipboard>
  )
})
