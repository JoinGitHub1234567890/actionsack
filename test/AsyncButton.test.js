import React from 'react'
import { render, fireEvent, wait } from 'react-testing-library'

import { AsyncButton } from '../src'

describe('<AsyncButton />', () => {
  test('should render children', async () => {
    const onClick = jest.fn(() => Promise.resolve('Success'))
    const { getByText, container } = render(
      <AsyncButton onClick={onClick}>
        {({ onClick, loading, data }) => (
          <button onClick={onClick}>{loading ? 'Loading' : data || 'Click'}</button>
        )}
      </AsyncButton>
    )

    expect(container.firstChild).toMatchSnapshot()

    fireEvent.click(getByText('Click'))

    expect(container.firstChild).toMatchSnapshot()

    await wait()

    expect(container.firstChild).toMatchSnapshot()
  })

  test('should render error when callback throws', async () => {
    const onClick = jest.fn(() => Promise.reject('Error'))
    const { getByText, container } = render(
      <AsyncButton onClick={onClick}>
        {({ onClick, error }) => <button onClick={onClick}>{error || 'Click'}</button>}
      </AsyncButton>
    )

    fireEvent.click(getByText('Click'))
    await wait()

    expect(container.firstChild).toMatchSnapshot()
  })
})
