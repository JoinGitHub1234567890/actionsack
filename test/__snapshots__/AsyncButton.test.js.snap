// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render children 1`] = `
<button>
  Click
</button>
`;

exports[`should render children 2`] = `
<button>
  Loading
</button>
`;

exports[`should render children 3`] = `
<button>
  Success
</button>
`;

exports[`should render error when callback throws 1`] = `
<button>
  Error
</button>
`;
